src/AcmeClient.php

```diff
@@ -11,79 +11,265 @@
+        Assert::nullOrString($email, 'registerAccount::$email expected a string or null. Got: %s');
+
+        $payload = [];
+        $payload['resource'] = ResourcesDirectory::NEW_REGISTRATION;
+        $payload['agreement'] = $agreement;
+
+        if ($email) {
+            $payload['contact'] = ['mailto:'.$email];
+        }
+
+        return $this->requestResource('POST', ResourcesDirectory::NEW_REGISTRATION, $payload);
     }
 
     /**
      * {@inheritdoc}
      */
-    protected function getCALicense()
+    public function requestAuthorization($domain)
     {
-        return $this->caLicense;
+        Assert::string($domain, 'requestChallenge::$domain expected a string. Got: %s');
+
+        $payload = [
+            'resource'   => ResourcesDirectory::NEW_AUTHORIZATION,
+            'identifier' => [
+                'type'  => 'dns',
+                'value' => $domain,
+            ],
+        ];
+
+        $response = $this->requestResource('POST', ResourcesDirectory::NEW_AUTHORIZATION, $payload);
+
+        if (!isset($response['challenges']) || !$response['challenges']) {
+            throw new HttpChallengeNotSupportedException();
+        }
+
+        $base64encoder = $this->httpClient->getBase64Encoder();
+        $keyParser = $this->httpClient->getKeyParser();
+        $accountKeyPair = $this->httpClient->getAccountKeyPair();
+
+        $parsedKey = $keyParser->parse($accountKeyPair->getPrivateKey());
+
+        foreach ($response['challenges'] as $challenge) {
+            if ('http-01' === $challenge['type']) {
+                $token = $challenge['token'];
+
+                $header = [
+                    // This order matters
+                    'e'   => $base64encoder->encode($parsedKey->getDetail('e')),
+                    'kty' => 'RSA',
+                    'n'   => $base64encoder->encode($parsedKey->getDetail('n')),
+                ];
+
+                $payload = $token.'.'.$base64encoder->encode(hash('sha256', json_encode($header), true));
+                $location = $this->httpClient->getLastLocation();
+
+                return new AuthorizationChallenge($domain, $challenge['uri'], $token, $payload, $location);
+            }
+        }
+
+        throw new HttpChallengeNotSupportedException();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function challengeAuthorization(AuthorizationChallenge $challenge, $timeout = 180)
+    {
+        Assert::integer($timeout, 'checkChallenge::$timeout expected an integer. Got: %s');
+
+        $payload = [
+            'resource'         => ResourcesDirectory::CHALLENGE,
+            'type'             => 'http-01',
+            'keyAuthorization' => $challenge->getPayload(),
+            'token'            => $challenge->getToken(),
+        ];
+
+        $response = $this->httpClient->signedRequest('POST', $challenge->getUrl(), $payload);
+
+        // Waiting loop
+        $waitingTime = 0;
+
+        while ($waitingTime < $timeout) {
```

Comment:
this is wrong, you compare a counter to a time. What if a SignedRequest (next line) spent 30 seconds on each loop ?...